code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/internal/EdgePersonalizationResponseHandler.java

```diff
@@ -261,92 +354,160 @@ void retrieveMessages(final List<Surface> surfaces, final Event event) {
-            feedRulesEngine.replaceRules(feedRules);
-        } else {
-            Log.debug(LOG_TAG, SELF_TAG, "Added %d rule(s) into the feed rules engine.", feedRules != null ? feedRules.size() : 0);
-            feedRulesEngine.addRules(feedRules);
+        // loop through propositions for this event and add them to existing proposition map by surface
+        for (final Proposition proposition : propositions) {
+            final Surface surface = Surface.fromUriString(proposition.getScope());
+            inProgressPropositions = MessagingUtils.updatePropositionMapForSurface(surface, proposition, inProgressPropositions);
         }
+    }
 
-        inboundMessages = feedRulesEngine.evaluate(edgeResponseEvent);
-        if (!MapUtils.isNullOrEmpty(inboundMessages)) {
-            updateInboundMessages(inboundMessages, requestedSurfaces);
-        }
+    private void beginRequestForSurfaces(final Event event, final List<Surface> surfaces) {
+        requestedSurfacesForEventId.put(event.getUniqueIdentifier(), surfaces);
+
+        // add the Edge request event to update propositions in the events queue.
+        serialWorkDispatcher.offer(event);
+    }
+
+    private void endRequestForEventId(final String eventId) {
+        // update in memory propositions
+        applyPropositionChangeForEventId(eventId);
+
+        // remove event from surfaces dictionary
+        requestedSurfacesForEventId.remove(eventId);
+
+        // clear pending propositions
+        inProgressPropositions.clear();
+    }
 
-        final Map<Surface, List<Proposition>> requestedPropositions = retrievePropositions(requestedSurfaces);
-        if (MapUtils.isNullOrEmpty(requestedPropositions)) {
-            Log.trace(MessagingConstants.LOG_TAG, SELF_TAG, "Not dispatching a notification event, personalization:decisions response does not contain propositions.");
+    private void applyPropositionChangeForEventId(final String eventId) {
+        // get the list of requested surfaces for this event
+        final List<Surface> requestedSurfaces = requestedSurfacesForEventId.get(eventId);
+        if (MessagingUtils.isNullOrEmpty(requestedSurfaces)) {
             return;
         }
 
-        // dispatch an event with the propositions received from the remote
-        final Map<String, Object> eventData = new HashMap<>();
-        final List<Map<String, Object>> convertedPropositions = new ArrayList<>();
-        for (final Map.Entry<Surface, List<Proposition>> propositionEntry : requestedPropositions.entrySet()) {
-            for (final Proposition proposition : propositionEntry.getValue()) {
-                convertedPropositions.add(proposition.toEventData());
+        final ParsedPropositions parsedPropositions = new ParsedPropositions(inProgressPropositions, requestedSurfaces, extensionApi);
+
+        // we need to preserve cache for any surfaces that were not a part of this request
```

Comment:
Do you mean "we need to preserve cache for any surfaces that were ~not~ a part of this request" ?